### ISSUE TITLE
Function: certificate signed by unknown authority "Crossplane"

### ISSUE THREAD

**Issue by withnale** [crossplane/crossplane#5456]
### What happened?

We are experiencing an intermittent problem with crossplane and the registration of the patch-and-transform function. We are constantly spinning down and spinning up clusters and we're now seeing this relatively regularly.

When the problem occurs we are seeing the following message appear in our logs and if we try to inspect the resource using `crossplane beta trace`:

> crossplane-6f458f9b86-plp42 universal-crossplane 2024-03-05T15:29:40Z	DEBUG	crossplane	Event(v1.ObjectReference{Kind:"XCompositeExample", Namespace:"", Name:"test05-qh6kd", UID:"e152b643-b2f7-4161-a893-fab54b1265d0", APIVersion:"platform.mydomain.com/v1alpha1", ResourceVersion:"110305", FieldPath:""}): type: 'Warning' reason: 'ComposeResources' cannot compose resources: cannot run Composition pipeline step "patch-and-transform": cannot run Function "crossplane-contrib-function-patch-and-transform": rpc error: code = Unavailable desc = last connection error: connection error: desc = "transport: authentication handshake failed: tls: failed to verify certificate: x509: certificate signed by unknown authority (possibly because of \"crypto/rsa: verification error\" while trying to verify candidate authority certificate \"Crossplane\")"

### How can we reproduce it?

As I mention, it's difficult to reproduce since I can spin up a cluster using exactly the same IaC branch and get a functioning deployment.

However at present it's very difficult to diagnose what the problem actually is. Other than the above message there is very little in the logs and my generate approach is to try to compare a broken cluster with a working cluster. I've looked at the TLS secrets associated with the pods and the certificates look fine. I've deleted the `upbound-system` namespace and let everything get recreated and the problem still persists.

If someone can shed some light on how to diagnose this problem further that would be great.

### What environment did it happen in?

Crossplane version:  0.14.4 and 0.15.0
Kubernetes version: v1.27.8 GKE
Function crossplane-contrib-function-patch-and-transform: 0.3.0

**Comment by phisco** [comment]
The root ca is deleted if you uninstall the crossplane helm chart, while functions' certificates are not, so if you then reinstall it, you end up with invalid certificates, a solution on our side could be to mark the root ca as not to be deleted, on your side, deleting the functions' secrets and letting Crossplane recreate them should solve the issue.

**Comment by withnale** [comment]
So these are the secret in question...

```
crossplane-contrib-function-patch-and-transform-tls-server   Opaque   3      156m
crossplane-root-ca                                           Opaque   2      156m
```

and although they exist when this problem occurs it's because the function secret needs to be signed by crossplane-root-ca and they have somehow got out of sync? Possibly that the root-ca has been changed after the function cert has been generated?

**Comment by withnale** [comment]
Okay so I can confirm that performing the following resolves this issue but it's not great.

```
kubectl delete  secret -n upbound-system crossplane-root-ca
kubectl delete  secret -n upbound-system -o name | grep -v crossplane-root-ca | xargs kubectl delete -n upbound-system
kubectl get pod -n upbound-system -o name | xargs kubectl delete -n upbound-system
```

Is there no way for the system to detect and resolve this automatically?

**Comment by bobh66** [comment]
If you delete the `upbound-system` namespace after you uninstall Crossplane it will clean up anything that gets left behind.

**Comment by withnale** [comment]
No. I think you misunderstand the question. We are deploying crossplane through gitops, so how do we ensure this doesn't occur. I don't want to manually have to intervene manually and ideally I don't want to have to put ordering into our end state definitions.

**Comment by Caomoji** [comment]
We are facing the same issue, both when setting up the cluster with ArgoCD and with a script. Even when the functions and providers and installed after Crossplane is fully ready it can still occur. We do not uninstall Crossplane, we install it once, then install the providers and functions. At this point, the `crossplane-root-ca` is populated by Crossplane's init container, so the certificate should be properly signed.

Cleaning the secrets and rolling out the pods does not necessarily solve this issue, sometime it does, sometime it doesn't; the proper order seems yet unclear. Could this be an issue with having multiple Crossplane replicas that don't sync correctly during the init?

**Comment by alexey-pankratyev** [comment]
I encountered the same problem when using function-environment-configs functions
`  Warning  ComposeResources   2s (x6 over 31s)    defined/compositeresourcedefinition.apiextensions.crossplane.io  cannot compose resources: cannot run Composition pipeline step "environmentConfigs": cannot run Function "function-environment-configs": rpc error: code = Unavailable desc = last connection error: connection error: desc = "transport: authentication handshake failed: tls: failed to verify certificate: x509: certificate signed by unknown authority (possibly because of \"crypto/rsa: verification error\" while trying to verify candidate authority certificate \"Crossplane\")"`

**Comment by turkenh** [comment]
Looking at this and I am wondering if we are getting an incorrect `NotFound` [here](https://github.com/crossplane/crossplane/blob/6b8545cc1caa6b5f57ccd7238dda7fbcb4c3d4af/internal/initializer/tls.go#L126) where we try to get the `crossplane-root-ca` secret from the controller runtime cache and regenerate the ca secret leaving previously generate function secrets out of sync.

I couldn't reproduce it yet, tried with multiple replicas, low cpu etc, but I wouldn't be surprised if 👆  is the root cause. If/once we could validate this (by somehow slowing down the controller runtime cache), fix should be easy I believe.

**Comment by truongnht** [comment]
@jbw976 is it fixed in v1.17, haven't seen #5892 mentioned in the release note of v1.17

**Comment by kubevalet** [comment]
I just upgraded helm chart to 1.17.2 and am still fighting with this issue :(

**Comment by aerfio** [comment]
Same issue encountered with the v1.18.2, suddenly our compositions stopped being reconciled (no status set on XRC) and there's a lot of events:

**Comment by jbw976** [comment]
@aerfio does the workaround described in https://github.com/crossplane/crossplane/issues/5456#issuecomment-1980992787 and shown in practice as an example in https://github.com/crossplane/crossplane/issues/5456#issuecomment-1988848053 work for you too? (note you'd probably need to change the namespace to `crossplane-system` instead).

Wondering if there's another path that results in this behavior that wasn't fixed by https://github.com/crossplane/crossplane/pull/5892, or if this is a different issue. Seeing if the workaround fixes it for you could be helpful 😁

**Comment by aerfio** [comment]
@jbw976  In our case we managed to fix this issue by deleting all the Crossplane secrets with the `tls-server` or `tls-client` suffix, we've seen in N times already, the biggest issue is that it happens randomly. We have almost the same situation as previously mentioned, we're dynamically creating k8s clusters. 

The issue also touches function and provider pods. I've observed that the IPs from the error message all point to the `konnectivity-agent` pods in kube-system namespace, it's an IBM cluster (IKS, k8s v1.31.2). 

We're investigating this issue right now, we'll get back with more info if we find anything, maybe it's just issue on IBM side

**Comment by franpog859** [comment]
It seems like something is replacing the `crossplane-root-ca` certificate. The other certificates are not signed by the Root CA and therefore we end up with the `TLS handshake error`
```
$ openssl x509 -text -noout -in temp/crossplane-root-ca.crt
Certificate:
    Data:
        Issuer: C=Earth, ST=Earth, L=Earth, O=Crossplane, CN=Crossplane
        Validity
            Not Before: Jan 15 10:31:32 2025 GMT
            Not After : Jan 15 10:31:32 2035 GMT
...

$ kubectl -n system logs crossplane-58f4667d9b-cx98t -c crossplane-init
{"level":"info","ts":"2025-01-14T11:51:07Z","logger":"crossplane","msg":"TLS CA secret is complete.","Step":"TLSCertificateGenerator"}
{"level":"info","ts":"2025-01-14T11:51:07Z","logger":"crossplane","msg":"Server certificates are empty or not complete, generating a new pair...","Step":"TLSCertificateGenerator","secret":"crossplane-tls-server"}
{"level":"info","ts":"2025-01-14T11:51:08Z","logger":"crossplane","msg":"Client certificates are empty or not complete, generating a new pair...","Step":"TLSCertificateGenerator","secret":"crossplane-tls-client"}
{"level":"info","ts":"2025-01-14T11:51:09Z","logger":"crossplane","msg":"Step has been completed","Name":"TLSCertificateGenerator"}
...
```

I'm checking it again on DigitalOcean

**Comment by franpog859** [comment]
On DigitalOcean I observed `TLS handshake error`s too. This time all the certificates are valid, none of them are replaced or not signed by the current `crossplane-root-ca`
```
$ kubectl -n system logs crossplane-7dcb555b7d-8qrkr -f
Defaulted container "crossplane" out of: crossplane, crossplane-init (init)
...
2025/01/16 12:29:04 http: TLS handshake error from 10.150.1.92:54158: EOF
2025/01/16 12:29:04 http: TLS handshake error from 10.150.0.198:35092: EOF
```

**Comment by hghtwr** [comment]
@jbw976 Just spitballing after hitting the same issue but is it possible that the fix implemented only works if the functions are passed to the helm chart using the parameters in the values file but not if they are deployed otherwise?

**Comment by semmet95** [comment]
Can this issue be reopened please? We are running into the same error with `v1.15.6`, and even though recreating `crossplane-system`, `crossplane-tls-client`, `crossplane-tls-server` works it is not very efficient when working with a lot of clusters in production.
I'm also concerned this secret recreation could have a side effect.

**Comment by prateekdegaons1991** [comment]
Hi Folks,
Can we reopen this issue,
even though we recreated the `certs crossplane-system`, `crossplane-tls-client`, `crossplane-tls-server`, the solution seems not working ! it still errors out !
or any other interim solution - we are stuck !

crossplane version: `v1.18.2`

error: 
```
Warning  ComposeResources  2m21s (x670 over 22h)  defined/compositeresourcedefinition.apiextensions.crossplane.io  cannot compose resources: cannot run Composition pipeline step "template parameter group": cannot run Function "function-go-templating": rpc error: code = DeadlineExceeded desc = latest balancer error: last connection error: connection error: desc = "transport: authentication handshake failed: tls: failed to verify certificate: x509: certificate signed by unknown authority (possibly because of \"crypto/rsa: verification error\" while trying to verify candidate authority certificate \"Crossplane\")"
```

**Comment by hghtwr** [comment]
@prateekdegaons1991 Afair, you have to recreate the function's tls secret, not the Crossplane core ones.

**Comment by prateekdegaons1991** [comment]
Well, this solution worked for us,

1. Scale down all the components in crossplane system 
2. Delete all the secrets 
3. Recreate these secrets crossplane core tls-client and tls-server 
```
kubectl create secret generic crossplane-tls-client --namespace=crossplane-system --type=Opaque
kubectl create secret generic crossplane-tls-server --namespace=crossplane-system --type=Opaque
```
4. scale up all the components

**Comment by jbw976** [comment]
Note that with https://github.com/crossplane/crossplane/pull/6917, which as been merged to `main` and will ship in v2.2, this scenario can be avoided more easily using `secrets.customAnnotations` when installing Crossplane's helm chart, as suggested by @ravilr:

```
secrets:
  # Instruct helm to not uninstall the resources
  # https://helm.sh/docs/howto/charts_tips_and_tricks/#tell-helm-not-to-uninstall-a-resource
  customAnnotations:
    helm.sh/resource-policy: keep
```

See the PR description for more details:

* https://github.com/crossplane/crossplane/pull/6917
